### ISSUE TITLE
Backfilling data results in metricDef writes to cassandra for every metric received

### ISSUE THREAD

**Issue by woodsaj** [raintank/metrictank#568]
The lastUpdate field stored with each metricDef in the index is the last timestamp seen for that metric.
So, when backfilling data, the lastUpdate field is set to a very old value. If the backlog of data is older then the update-interval, then every metric received will be compared against time.Now() and treated as old and needing to be updated.

The quick fix is to just set lastUpdate to time.Now().  But a more thorough fix would be to keep 2 timestamps.  the existing "lastUpdate" to represent the last time a metric was received and a LastSave to represent the last time the metric was written to cassandra.

**Comment by Dieterbe** [comment]
I went through all occurences of `/lastupdate/i` in the code, to confirm what the intended uses are.
Specifically I wanted to find out why do we want to track last timestamp of a datapoint (hereafter called data time) vs last timestamp of the wall clock at the time a point was seen (hereafter called clock time)

* index filtering in render request path if last data timestamp < from. (data time)
* index pruning. old lastupdate -> remove from index. this should be clock time
* updating the def in the index when it's already there. compares against wall-clock (but frankly i don't see a use for re-saving a def to the index if it's already there)
* mt-index-cat uses LastUpdate to not print defs for "old" metrics, it's unclear what old means here. both data time and clock time could be useful.

**Comment by Dieterbe** [comment]
as for where LastUpdate properties are being set. there is one place : `MemoryIdx.AddOrUpdate -> lastUpdate = data.Time` (and also when loading a cassandra index into memory)

**Comment by Dieterbe** [comment]
and btw, there's also AggMetric.lastWrite which is set to wallclock when a point arrives. the max-stale options compare the values against wall-clock when deciding whether to keep the data in the ringbuffers. (lastWrite is not synced to the index, nor should it)

**Comment by Dieterbe** [comment]
> and a LastSave to represent the last time the metric was written to cassandra.

what would be the use of that?

**Comment by woodsaj** [comment]
> what would be the use of that?

in cassandraIdx, the only reason we look at the lastUpdate is because it used to represent the last time the def was written to cassandra, and if it is too old we need to write an update to cassandra.

in memoryIdx, we use LastUpdate as the last time a metric was seen.  We then use that for filtering results in Find() queries.

**Comment by Dieterbe** [comment]
I just realized something. (maybe it's obvious to you already) : 
to make index filtering work reliably for recently started nodes that are queried for data that lives in the partitions they "own" but have not ingested points for, and hence have to rely on the defs they loaded from cassandra, we have to make sure:
1) offset seek back when they replay data at startup > update-interval being used for those partitions. ideally a lot more to account for fluctuations in queue times. in other words update-interval should be no larger than half the window of data that will replayed on startup.
2) while normally, we will always have some nodes with low priority (caught up, and hence with guaranteed uptodate lastUpdate fields, if we followed point 1) we could use max-priority to set an upper bound on the risk of getting an out-of-date lastupdate from cassandra we want to take. but this seems a bit extreme. probably better to mark the node as ready in such case and be subject to the filtering potentially being a bit too eager. better that then having no nodes available for reads.
